[sniff] Multi line function declare

dcolt opened the pull request:
Re-added the rule to allow functions with empty body to be declared on a single line with an empty body.

dcolt committed:
[sniff] Multi line function declare

sunkan commented:
Can you add some more test cases

Verify that return type don't interfer
```
<?php declare(strict_types=1);

class Test
{
	public function A(): SomeClass { }
}
```

Some with arguments

```
<?php declare(strict_types=1);

class Test
{
	public function __construct(
		private string $value,
	) { }
}
```

Maybe some more if you can think about it

dcolt committed:
[tests] Added more testcases for MultiLineFunctionDeclationSniff

dcolt commented:
@sunkan I've added a few more, and also some "cursed" valid ones, which I suspect we might want to update the rules for.

dcolt committed:
[tests] Added edge cases

Seems like something that shouldn't be allowed

dcolt commented on tests/Sniffs/Functions/data/MultiLineFunctionDeclarationSniff.OK.php:
```diff
@@ -0,0 +1,16 @@
+
+class Test
+{
+	public function __construct(private string $str){}
```
I would suggest we should add a rule that forces the opening bracket to have exactly one space between itself and the previous non whitespace token?

bix0r replied:
I agree. I would say that only D and E are "OK" in this file.

dcolt commented on tests/Sniffs/Functions/data/MultiLineFunctionDeclarationSniff.OK.php:
```diff
@@ -0,0 +1,16 @@
+
+	public function A() {}
+
+	public function B(): string{}
```
Same white space issue here

dcolt commented on tests/Sniffs/Functions/data/MultiLineFunctionDeclarationSniff.OK.php:
```diff
@@ -0,0 +1,16 @@
+
+	public function B(): string{}
+
+	public function C(): string|null  {}
```
Extra whitespace.

bix0r commented on src/Stefna/Sniffs/Functions/MultiLineFunctionDeclarationSniff.php:
```diff
@@ -0,0 +1,74 @@
+
+		parent::processSingleLineDeclaration($phpcsFile, $stackPtr, $tokens);
+	}
+	/**
```
Seems like this file needs formatting. Also, `@param array` is useless, since you don't have any hints of what this array could be

bix0r reviewed (changes requested):
This is a "multi-line-function-decleration" sniff, right? Where are the tests for multi-line stuff?

dcolt commented:
@bix0r 
> This is a "multi-line-function-decleration" sniff, right? Where are the tests for multi-line stuff?

I believe it's a bit of a misleading name.
It's extending the Squiz MultilineFunctionDeclarationSniff, but our stuff is explicitly for the "OneLineMethod", which requires the line scope start and end to be on the same line.

And adding tests for what comes from the base, seems to be out of the scope for the tests?

dcolt committed:
[sniff,MultiLineFunctionDeclatration] Add check for whitespace before `{`

dcolt committed:
[test] refactor TestCase

- Test files now live under `data/<SniffName>`
- `fixed.php` files are automatically verified to not have any errors
- Removed the need to pass around the report file.
- When assertingAllFixedInfile also verified all erros have been
  checked.

dcolt committed:
[test] Remove property constructor.

It will not be allowed in combination with future sniffs.

dcolt committed:
[tests] Resolve rebase issues

dcolt committed:
[docs] Add readme entry for MultiLineFunctionDeclarationSniff
